src/lib.rs

```diff
@@ -1,171 +1,278 @@
-        .wrapping_add(1442695040888963407)
-        .wrapping_add(rdtsc());
-    seed.wrapping_mul(6364136223846793005)
-        .wrapping_add(1442695040888963407)
-}
+    fn send_time_ms(&self, key: &str, interval_ms: u64) {
+        let value = &interval_ms.to_string();
+        self.send( &[key, ":", value, "|ms|@", &self.str_rate] )
+    }
 
-fn pcg32() -> u32 {
-    thread_local! {
-        static PCG32_STATE: RefCell<u64> = RefCell::new(seed());
+    /// Concatenate text parts into a single buffer and send it over UDP
+    fn send(&self, strings: &[&str]) {
+        let mut str = String::with_capacity(MAX_UDP_PAYLOAD);
+        str.push_str(&self.prefix);
+        for s in strings { str.push_str(s); }
+        str.push_str(&self.str_rate);
+        match self.udp_socket.send(str.as_bytes()) {
+            Ok(_) => {}, // TODO count packets sent for batch reporting
+            _ => {}// TODO count send errors for batch reporting
+        }
     }
 
-    PCG32_STATE.with(|state| {
-        let oldstate: u64 = *state.borrow();
-        // XXX could generate the increment from the thread ID
-        *state.borrow_mut() = oldstate.wrapping_mul(6364136223846793005)
-            .wrapping_add(1442695040888963407);
-        ((((oldstate >> 18) ^ oldstate) >> 27) as u32)
-            .rotate_right((oldstate >> 59) as u32)
-    })
 }
 
-pub fn within_rate(r: u32) -> bool {
-    pcg32() > r
+/// Convert a floating point sampling rate to an integer so that a fast integer RNG can be used
+/// Float rate range is between 1.0 (send 100% of the samples) and 0.0 (_no_ samples taken)
+/// .    | float rate | int rate | percentage
+/// ---- | ---------- | -------- | ----
+/// all  | 1.0        | 0x0      | 100%
+/// none | 0.0        | 0xFFFFFFFF | 0%
+fn to_int_rate(float_rate: f64) -> u32 {
+    assert!(float_rate <= 1.0 && float_rate >= 0.0);
+    ((1.0 - float_rate) * ::std::u32::MAX as f64) as u32
 }
 
+fn accept_sample(int_rate: u32) -> bool {
+    pcg32::random() > int_rate
+}
 
-// XXX we need to redo this as a syntax extension that stringifies all
-// the constant arguments so we don't need to format them.  This is
-// particularly important for rate.
+/// A convenience macro to wrap a block or an expression with a start / stop timer.
+/// Elapsed time is sent to the supplied statsd client after the computation has been performed.
+/// Expression result (if any) is transparently returned.
 #[macro_export]
-macro_rules! statsd {
-    (increment, $key:expr, $count:expr, $rate:expr) => ({
-        if unsafe { ::std::intrinsics::unlikely($crate::within_rate(((1.0 - $rate) * ::std::u32::MAX as f64) as u32)) } {
-            let count = format!("{}", $count);
-            let rate = format!("{}", $rate);
-            $crate::send(&[$key, b":", count.as_bytes(), b"|c", b"|@", rate.as_bytes()])
+macro_rules! time {
+    ($client: expr, $key: expr, $body: block) => (
+        let start_time = time::precise_time_ns();
```

Comment:
You probably want to write:
```rust
let start_time = ::time::precise_time_ns();
```
so that you don't need to import the time root module in every module you use this `time!` macro.

That requires that every crate using this macro will need to use the `time` crate.
An alternative would be to reexport this `precise_time_ns` function from this crate and write:
```rust
let start_time = $crate::precise_time_ns();
```